File: src/Wakeel.Crypto/Identity/DeviceIdentity.cs
using System.Text;
using NSec.Cryptography;
using NKey = NSec.Cryptography.Key;
using NPublicKey = NSec.Cryptography.PublicKey;

namespace Wakeel.Crypto;

/// <summary>
/// A device key pair set: Ed25519 for signing everything the device produces and
/// X25519 for agreeing session keys and opening sealed boxes.
/// The organisation key and the office computer key use the same type.
/// </summary>
public sealed class DeviceIdentity : IDisposable
{
    public const int PublicKeySize = 32;
    public const int SignatureSize = 64;

    private const string AgreeLabel = "wakeel.agree";
    private const string SealLabel = "wakeel.seal";

    private readonly NKey _signingKey;
    private readonly NKey _agreementKey;
    private bool _disposed;

    private DeviceIdentity(NKey signingKey, NKey agreementKey)
    {
        _signingKey = signingKey;
        _agreementKey = agreementKey;
        SigningPublicKey = signingKey.PublicKey.Export(KeyBlobFormat.RawPublicKey);
        AgreementPublicKey = agreementKey.PublicKey.Export(KeyBlobFormat.RawPublicKey);
    }

    /// <summary>Raw Ed25519 public key.</summary>
    public byte[] SigningPublicKey { get; }

    /// <summary>Raw X25519 public key.</summary>
    public byte[] AgreementPublicKey { get; }

    /// <summary>Ed25519 public key in the base64url form stored in certificates.</summary>
    public string SigningPublicKeyText => Base64Url.Encode(SigningPublicKey);

    /// <summary>X25519 public key in the base64url form stored in certificates.</summary>
    public string AgreementPublicKeyText => Base64Url.Encode(AgreementPublicKey);

    public static DeviceIdentity Generate() =>
        Import(new DeviceSeeds(RandomBytes.Next(DeviceSeeds.SeedSize), RandomBytes.Next(DeviceSeeds.SeedSize)));

    public static DeviceIdentity Import(DeviceSeeds seeds)
    {
        ArgumentNullException.ThrowIfNull(seeds);
        seeds.Validate();

        var creation = new KeyCreationParameters { ExportPolicy = KeyExportPolicies.AllowPlaintextExport };
        NKey? signing = null;
        NKey? agreement = null;
        try
        {
            signing = NKey.Import(SignatureAlgorithm.Ed25519, seeds.SigningSeed, KeyBlobFormat.RawPrivateKey, in creation);
            agreement = NKey.Import(KeyAgreementAlgorithm.X25519, seeds.AgreementSeed, KeyBlobFormat.RawPrivateKey, in creation);
            return new DeviceIdentity(signing, agreement);
        }
        catch (FormatException exception)
        {
            signing?.Dispose();
            agreement?.Dispose();
            throw new CryptoException(ErrorCode.Corrupt, "The device seeds could not be imported.", exception);
        }
    }

    /// <summary>Exports the private seeds so they can be wrapped and stored.</summary>
    public DeviceSeeds Export()
    {
        ThrowIfDisposed();
        return new DeviceSeeds(
            _signingKey.Export(KeyBlobFormat.RawPrivateKey),
            _agreementKey.Export(KeyBlobFormat.RawPrivateKey));
    }

    public byte[] Sign(ReadOnlySpan<byte> data)
    {
        ThrowIfDisposed();
        return SignatureAlgorithm.Ed25519.Sign(_signingKey, data);
    }

    /// <summary>Signs and returns the base64url text stored next to a signed body.</summary>
    public string SignText(ReadOnlySpan<byte> data) => Base64Url.Encode(Sign(data));

    public static bool Verify(ReadOnlySpan<byte> signingPublicKey, ReadOnlySpan<byte> data, ReadOnlySpan<byte> signature)
    {
        if (signingPublicKey.Length != PublicKeySize || signature.Length != SignatureSize)
        {
            return false;
        }

        NPublicKey publicKey;
        try
        {
            publicKey = NPublicKey.Import(SignatureAlgorithm.Ed25519, signingPublicKey, KeyBlobFormat.RawPublicKey);
        }
        catch (FormatException)
        {
            return false;
        }

        return SignatureAlgorithm.Ed25519.Verify(publicKey, data, signature);
    }

    public static bool Verify(string signingPublicKeyText, ReadOnlySpan<byte> data, string signatureText)
    {
        if (!Base64Url.TryDecode(signingPublicKeyText, out var publicKey)
            || !Base64Url.TryDecode(signatureText, out var signature))
        {
            return false;
        }

        return Verify(publicKey, data, signature);
    }

    /// <summary>
    /// Derives a shared session key with a peer. Both sides get the same bytes because the
    /// salt orders the two public keys, so neither side needs to know who started.
    /// </summary>
    public byte[] Agree(ReadOnlySpan<byte> peerAgreementPublicKey, string context, int length = 32)
    {
        ThrowIfDisposed();
        var salt = OrderedSalt(AgreementPublicKey, peerAgreementPublicKey);
        var info = Encoding.UTF8.GetBytes($"{AgreeLabel}|{context}");
        return Derive(_agreementKey, peerAgreementPublicKey, salt, info, length);
    }

    /// <summary>
    /// Sealed box: an ephemeral X25519 key pair encrypts a payload to a recipient public key.
    /// The ephemeral public key is the first thirty two bytes of the result.
    /// </summary>
    public static byte[] SealFor(ReadOnlySpan<byte> recipientAgreementPublicKey, ReadOnlySpan<byte> payload, string context)
    {
        if (recipientAgreementPublicKey.Length != PublicKeySize)
        {
            throw new CryptoException(ErrorCode.Corrupt, "The recipient public key has the wrong length.");
        }

        var creation = new KeyCreationParameters { ExportPolicy = KeyExportPolicies.AllowPlaintextExport };
        using var ephemeral = NKey.Create(KeyAgreementAlgorithm.X25519, in creation);
        var ephemeralPublic = ephemeral.PublicKey.Export(KeyBlobFormat.RawPublicKey);

        var salt = Concat(ephemeralPublic, recipientAgreementPublicKey);
        var info = Encoding.UTF8.GetBytes($"{SealLabel}|{context}");
        var key = Derive(ephemeral, recipientAgreementPublicKey, salt, info, Aead.KeySize);

        var sealedPayload = Aead.Encrypt(key, payload, salt);
        var result = new byte[ephemeralPublic.Length + sealedPayload.Length];
        ephemeralPublic.CopyTo(result, 0);
        sealedPayload.CopyTo(result, ephemeralPublic.Length);
        return result;
    }

    /// <summary>Opens a sealed box addressed to this identity's X25519 public key.</summary>
    public byte[] Open(ReadOnlySpan<byte> sealedBox, string context)
    {
        ThrowIfDisposed();
        if (sealedBox.Length <= PublicKeySize)
        {
            throw new CryptoException(ErrorCode.Corrupt, "The sealed block is too short to be valid.");
        }

        var ephemeralPublic = sealedBox[..PublicKeySize];
        var body = sealedBox[PublicKeySize..];
        var salt = Concat(ephemeralPublic, AgreementPublicKey);
        var info = Encoding.UTF8.GetBytes($"{SealLabel}|{context}");
        var key = Derive(_agreementKey, ephemeralPublic, salt, info, Aead.KeySize);
        return Aead.Decrypt(key, body, salt);
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        _signingKey.Dispose();
        _agreementKey.Dispose();
    }

    private static byte[] Derive(
        NKey privateKey,
        ReadOnlySpan<byte> peerPublicKey,
        ReadOnlySpan<byte> salt,
        ReadOnlySpan<byte> info,
        int length)
    {
        if (peerPublicKey.Length != PublicKeySize)
        {
            throw new CryptoException(ErrorCode.Corrupt, "The peer public key has the wrong length.");
        }

        NPublicKey peer;
        try
        {
            peer = NPublicKey.Import(KeyAgreementAlgorithm.X25519, peerPublicKey, KeyBlobFormat.RawPublicKey);
        }
        catch (FormatException exception)
        {
            throw new CryptoException(ErrorCode.Corrupt, "The peer public key could not be read.", exception);
        }

        using var shared = KeyAgreementAlgorithm.X25519.Agree(privateKey, peer)
            ?? throw new CryptoException(ErrorCode.Corrupt, "The two devices could not agree on a shared key.");

        return KeyDerivationAlgorithm.HkdfSha256.DeriveBytes(shared, salt, info, length);
    }

    private static byte[] OrderedSalt(ReadOnlySpan<byte> first, ReadOnlySpan<byte> second) =>
        first.SequenceCompareTo(second) <= 0 ? Concat(first, second) : Concat(second, first);

    private static byte[] Concat(ReadOnlySpan<byte> first, ReadOnlySpan<byte> second)
    {
        var buffer = new byte[first.Length + second.Length];
        first.CopyTo(buffer);
        second.CopyTo(buffer.AsSpan(first.Length));
        return buffer;
    }

    private void ThrowIfDisposed() => ObjectDisposedException.ThrowIf(_disposed, this);
}

File: android/core/src/main/kotlin/ps/wakeel/phone/core/crypto/DeviceIdentity.kt
package ps.wakeel.phone.core.crypto

import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer

/**
 * Every signature the product makes is taken over a labelled buffer, so a signature produced
 * for one kind of structure can never be read as a signature over another. The labels are the
 * ones in `Wakeel.Crypto/Json/DomainSeparation.cs`, character for character.
 */
object DomainSeparation {
    const val CERTIFICATE = "wakeel.cert|v1"
    const val REVOCATION_LIST = "wakeel.revocations|v1"
    const val PAIRING_REQUEST = "wakeel.pairing.request|v1"
    const val PAIRING_ACCEPT = "wakeel.pairing.accept|v1"
    const val CONTAINER = "wakeel.container.sig|v1"

    /** Returns `label || body`, the exact bytes a signature covers. */
    fun wrap(label: String, body: ByteArray): ByteArray {
        val prefix = label.toByteArray(Charsets.UTF_8)
        val buffer = ByteArray(prefix.size + body.size)
        prefix.copyInto(buffer, 0)
        body.copyInto(buffer, prefix.size)
        return buffer
    }
}

/**
 * The two private seeds of a device. They are generated on the phone, never leave it, and are
 * kept wrapped by an Android Keystore key — see [KeyVault].
 */
class DeviceSeeds(signingSeed: ByteArray, agreementSeed: ByteArray) {
    companion object {
        const val SEED_SIZE = 32

        fun generate(): DeviceSeeds = DeviceSeeds(RandomBytes.next(SEED_SIZE), RandomBytes.next(SEED_SIZE))
    }

    val signingSeed: ByteArray = signingSeed.copyOf()
    val agreementSeed: ByteArray = agreementSeed.copyOf()

    init {
        if (this.signingSeed.size != SEED_SIZE || this.agreementSeed.size != SEED_SIZE) {
            throw CryptoException(CryptoErrorCode.Corrupt, "a device seed must be thirty two bytes")
        }
    }

    /** The two seeds, one after the other — the form the key vault wraps. */
    fun toBytes(): ByteArray = signingSeed + agreementSeed

    fun clear() {
        signingSeed.fill(0)
        agreementSeed.fill(0)
    }
}

fun deviceSeedsFromBytes(bytes: ByteArray): DeviceSeeds {
    if (bytes.size != DeviceSeeds.SEED_SIZE * 2) {
        throw CryptoException(CryptoErrorCode.Corrupt, "the stored device seeds are the wrong length")
    }
    return DeviceSeeds(
        bytes.copyOfRange(0, DeviceSeeds.SEED_SIZE),
        bytes.copyOfRange(DeviceSeeds.SEED_SIZE, bytes.size),
    )
}

/**
 * A device key pair set: Ed25519 for signing everything the device produces and X25519 for
 * agreeing session keys and opening sealed boxes. Byte compatible with
 * `Wakeel.Crypto/Identity/DeviceIdentity.cs`.
 */
class DeviceIdentity private constructor(private val seeds: DeviceSeeds) {

    companion object {
        const val PUBLIC_KEY_SIZE = 32
        const val SIGNATURE_SIZE = 64

        private const val AGREE_LABEL = "wakeel.agree"
        private const val SEAL_LABEL = "wakeel.seal"

        fun generate(): DeviceIdentity = DeviceIdentity(DeviceSeeds.generate())

        fun import(seeds: DeviceSeeds): DeviceIdentity = DeviceIdentity(seeds)

        fun verify(signingPublicKey: ByteArray, data: ByteArray, signature: ByteArray): Boolean {
            if (signingPublicKey.size != PUBLIC_KEY_SIZE || signature.size != SIGNATURE_SIZE) return false
            return try {
                val signer = Ed25519Signer()
                signer.init(false, Ed25519PublicKeyParameters(signingPublicKey, 0))
                signer.update(data, 0, data.size)
                signer.verifySignature(signature)
            } catch (_: IllegalArgumentException) {
                false
            } catch (_: RuntimeException) {
                false
            }
        }

        fun verify(signingPublicKeyText: String?, data: ByteArray, signatureText: String?): Boolean {
            val key = Base64Url.tryDecode(signingPublicKeyText) ?: return false
            val signature = Base64Url.tryDecode(signatureText) ?: return false
            return verify(key, data, signature)
        }

        /**
         * Sealed box: an ephemeral X25519 key pair encrypts a payload to a recipient public key.
         * The ephemeral public key is the first thirty two bytes of the result.
         */
        fun sealFor(recipientAgreementPublicKey: ByteArray, payload: ByteArray, context: String): ByteArray {
            if (recipientAgreementPublicKey.size != PUBLIC_KEY_SIZE) {
                throw CryptoException(CryptoErrorCode.Corrupt, "the recipient public key has the wrong length")
            }
            val ephemeralSeed = RandomBytes.next(DeviceSeeds.SEED_SIZE)
            val ephemeralPrivate = X25519PrivateKeyParameters(ephemeralSeed, 0)
            val ephemeralPublic = ephemeralPrivate.generatePublicKey().encoded

            val salt = ephemeralPublic + recipientAgreementPublicKey
            val info = "$SEAL_LABEL|$context".toByteArray(Charsets.UTF_8)
            val shared = agree(ephemeralPrivate, recipientAgreementPublicKey)
            val key = Hkdf.deriveKey(shared, Aead.KEY_SIZE, salt, info)

            val sealed = Aead.encrypt(key, payload, salt)
            return ephemeralPublic + sealed
        }

        internal fun agree(privateKey: X25519PrivateKeyParameters, peerPublicKey: ByteArray): ByteArray {
            if (peerPublicKey.size != PUBLIC_KEY_SIZE) {
                throw CryptoException(CryptoErrorCode.Corrupt, "the peer public key has the wrong length")
            }
            val agreement = X25519Agreement()
            agreement.init(privateKey)
            val shared = ByteArray(agreement.agreementSize)
            try {
                agreement.calculateAgreement(X25519PublicKeyParameters(peerPublicKey, 0), shared, 0)
            } catch (exception: RuntimeException) {
                throw CryptoException(CryptoErrorCode.Corrupt, "the two devices could not agree on a key", exception)
            }
            return shared
        }
    }

    private val signingPrivate = Ed25519PrivateKeyParameters(seeds.signingSeed, 0)
    private val agreementPrivate = X25519PrivateKeyParameters(seeds.agreementSeed, 0)

    /** Raw Ed25519 public key. */
    val signingPublicKey: ByteArray = signingPrivate.generatePublicKey().encoded

    /** Raw X25519 public key. */
    val agreementPublicKey: ByteArray = agreementPrivate.generatePublicKey().encoded

    val signingPublicKeyText: String get() = Base64Url.encode(signingPublicKey)

    val agreementPublicKeyText: String get() = Base64Url.encode(agreementPublicKey)

    /** The seeds, so the caller can wrap and store them; nothing else may read them. */
    fun exportSeeds(): DeviceSeeds = seeds

    fun sign(data: ByteArray): ByteArray {
        val signer = Ed25519Signer()
        signer.init(true, signingPrivate)
        signer.update(data, 0, data.size)
        return signer.generateSignature()
    }

    fun signText(data: ByteArray): String = Base64Url.encode(sign(data))

    /**
     * Derives a shared key with a peer. Both sides get the same bytes because the salt orders
     * the two public keys, so neither side has to know who started.
     */
    fun agreeWith(peerAgreementPublicKey: ByteArray, context: String, length: Int = 32): ByteArray {
        val salt = orderedSalt(agreementPublicKey, peerAgreementPublicKey)
        val info = "$AGREE_LABEL|$context".toByteArray(Charsets.UTF_8)
        val shared = agree(agreementPrivate, peerAgreementPublicKey)
        return Hkdf.deriveKey(shared, length, salt, info)
    }

    /** Opens a sealed box addressed to this identity's X25519 public key. */
    fun open(sealedBox: ByteArray, context: String): ByteArray {
        if (sealedBox.size <= PUBLIC_KEY_SIZE) {
            throw CryptoException(CryptoErrorCode.Corrupt, "the sealed block is too short")
        }
        val ephemeralPublic = sealedBox.copyOfRange(0, PUBLIC_KEY_SIZE)
        val body = sealedBox.copyOfRange(PUBLIC_KEY_SIZE, sealedBox.size)
        val salt = ephemeralPublic + agreementPublicKey
        val info = "$SEAL_LABEL|$context".toByteArray(Charsets.UTF_8)
        val shared = agree(agreementPrivate, ephemeralPublic)
        val key = Hkdf.deriveKey(shared, Aead.KEY_SIZE, salt, info)
        return Aead.decrypt(key, body, salt)
    }

    private fun orderedSalt(first: ByteArray, second: ByteArray): ByteArray =
        if (compareUnsigned(first, second) <= 0) first + second else second + first

    /**
     * Bytes compare as unsigned numbers here, because that is what .NET's span comparison does;
     * Kotlin's own byte comparison is signed and would order the two keys the other way round
     * for any key whose first differing byte is above 0x7F.
     */
    private fun compareUnsigned(left: ByteArray, right: ByteArray): Int {
        val shared = minOf(left.size, right.size)
        for (index in 0 until shared) {
            val difference = (left[index].toInt() and 0xFF) - (right[index].toInt() and 0xFF)
            if (difference != 0) return difference
        }
        return left.size - right.size
    }
}
